pkg/gofr/datasource/dbresolver/resolver.go

```diff
@@ -0,0 +1,504 @@
+	for _, opt := range opts {
+		opt(r)
+	}
+
+	// Initialize metrics and start background tasks.
+	r.initializeMetrics()
+	r.startBackgroundTasks()
+
+	if r.logger != nil {
+		r.logger.Logf("DB Resolver initialized with %d replicas using circuit breakers", len(replicas))
+	}
+
+	return r
+}
+
+// initializeMetrics sets up metrics following GoFr patterns.
+func (r *Resolver) initializeMetrics() {
+	if r.metrics == nil {
+		return
+	}
+
+	// Histogram for query response times
+	buckets := []float64{0.05, 0.1, 0.2, 0.5, 1.0, 2.0, 5.0, 10.0}
+	r.metrics.NewHistogram("dbresolver_query_duration", "Response time of DB resolver operations in microseconds", buckets...)
+
+	// Gauges for operation tracking
+	r.metrics.NewGauge("dbresolver_primary_reads", "Total reads routed to primary")
+	r.metrics.NewGauge("dbresolver_primary_writes", "Total writes routed to primary")
+	r.metrics.NewGauge("dbresolver_replica_reads", "Total reads routed to replicas")
+	r.metrics.NewGauge("dbresolver_fallbacks", "Total fallbacks to primary")
+	r.metrics.NewGauge("dbresolver_failures", "Total replica failures")
+}
+
+// startBackgroundTasks starts minimal background processing.
+func (r *Resolver) startBackgroundTasks() {
+	r.wg.Add(1)
+	go r.backgroundProcessor()
+}
+
+// backgroundProcessor handles metrics collection with reduced frequency.
+func (r *Resolver) backgroundProcessor() {
+	defer r.wg.Done()
+
+	ticker := time.NewTicker(time.Duration(defaultTimeoutSec) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-r.stopChan:
+			return
+		case <-ticker.C:
+			r.updateMetrics()
+		}
+	}
+}
+
+// updateMetrics updates gauge metrics.
+func (r *Resolver) updateMetrics() {
+	if r.metrics == nil {
+		return
+	}
+
+	r.metrics.SetGauge("dbresolver_primary_reads", float64(r.stats.primaryReads.Load()))
+	r.metrics.SetGauge("dbresolver_primary_writes", float64(r.stats.primaryWrites.Load()))
+	r.metrics.SetGauge("dbresolver_replica_reads", float64(r.stats.replicaReads.Load()))
+	r.metrics.SetGauge("dbresolver_fallbacks", float64(r.stats.primaryFallbacks.Load()))
+	r.metrics.SetGauge("dbresolver_failures", float64(r.stats.replicaFailures.Load()))
+}
+
+// Fast query classification with optimized string operations.
+func (r *Resolver) isReadQuery(query string) bool {
+	// Fast path: check first few characters for common patterns
+	if len(query) < minQueryLength {
+		return false
+	}
+
+	// Trim whitespace and get first word
+	trimmed := strings.TrimLeft(query, " \t\n\r")
+	if len(trimmed) < minQueryLength {
+		return false
+	}
+
+	// Fast string comparison for common cases
+	firstSix := strings.ToUpper(trimmed[:6])
+	switch firstSix {
+	case "SELECT", "SHOW  ", "DESCRI", "EXPLAI":
+		return true
+	}
+
+	// Check cache for edge cases
+	if cached, exists := r.queryCache.get(query); exists {
+		return cached
+	}
+
+	// Fallback to regex for complex queries
+	isRead := readQueryRegex.MatchString(query)
+
+	r.queryCache.set(query, isRead)
+
+	return isRead
+}
```

Comment:
I reduced size to 4 because no one will use show followed by 2 spaces, especially after you trimmed.

```suggestion
	const minQueryLength = 4

	// Fast path: check first few characters for common patterns
	// Trim whitespace and get first word
	trimmed := strings.TrimLeft(query, " \t\n\r")
	if len(trimmed) < minQueryLength {
		return false
	}

	// Fast string comparison for common cases
	firstSix := strings.ToUpper(trimmed[:minQueryLength])
	switch firstSix {
	case "SELE", "SHOW", "DESC", "EXPL":
		return true
	}

	// Check cache for edge cases
	if cached, exists := r.queryCache.get(query); exists {
		return cached
	}

	// Fallback to regex for complex queries
	isRead := readQueryRegex.MatchString(query)

	r.queryCache.set(query, isRead)

	return isRead
}
```